In [2]:
# import torch
# print(f"PyTorch Version: {torch.__version__}")
# print(f"CUDA Version: {torch.version.cuda}")

In [3]:
# import torch
# try:
#     import torch_sparse
#     from torch_sparse import SparseTensor
#     print("✅ 安装成功！环境配置修复完成。")
# except OSError as e:
#     print("❌ 安装失败或版本仍不匹配。请尝试下方的【方案二】。")
#     print(e)
# except ImportError as e:
#     print("❌ 导入失败，请尝试下方的【方案二】。")
#     print(e)

In [5]:
!nvidia-smi

Thu Jan 15 20:17:30 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.195.03             Driver Version: 570.195.03     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4090        Off |   00000000:16:00.0 Off |                  Off |
| 31%   33C    P8             23W /  450W |     805MiB /  24564MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [5]:
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "2" # 假设我们要用2号卡

import torch
print(torch.cuda.is_available()) # 这里必须输出 True
print(torch.cuda.device_count()) # 这里应该输出 1
print(torch.cuda.current_device()) # 这里应该输出 0
print(torch.cuda.get_device_name(0)) # 这里应该输出 RTX 4090

True
1
0
NVIDIA GeForce RTX 4090


In [6]:
!python Step1_ConstructCellularSpatialGraphs.py

2026-01-14 21:02:18
Constructing topology structures of KNN graphs...
This is image-0
This is image-1
This is image-2
This is image-3
This is image-4
This is image-5
This is image-6
This is image-7
This is image-8
This is image-9
This is image-10
This is image-11
This is image-12
This is image-13
This is image-14
This is image-15
This is image-16
This is image-17
This is image-18
This is image-19
This is image-20
This is image-21
This is image-22
This is image-23
This is image-24
This is image-25
This is image-26
This is image-27
This is image-28
This is image-29
This is image-30
This is image-31
This is image-32
This is image-33
All topology structures have been generated!
2026-01-14 21:02:53
Generating node attribute matrices of KNN graphs...
This is image-0
This is image-1
This is image-2
This is image-3
This is image-4
This is image-5
This is image-6
This is image-7
This is image-8
This is image-9
This is image-10
This is image-11
This is image-12
This is image-13
This is image-14


2026-01-14 16:36:06
Constructing topology structures of KNN graphs...
This is image-0
This is image-1
This is image-2
This is image-3
This is image-4
This is image-5
This is image-6
This is image-7
This is image-8
This is image-9
This is image-10
This is image-11
This is image-12
This is image-13
This is image-14
This is image-15
This is image-16
This is image-17
This is image-18
This is image-19
This is image-20
This is image-21
This is image-22
...
Processing...
Done!
Step1 done!
2026-01-14 16:36:46

In [7]:
!python Step2_TCNLearning_Supervised.py

/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step2_TCNLearning_Supervised.py:39: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  self.data, self.slices = tor

2026-01-14 16:36:47
This is time: 01
2026-01-14 16:36:47
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step2_TCNLearning_Supervised.py:189: DeprecationWarning: Sampling from a set deprecated
since Python 3.9 and will be removed in a subsequent version.
  test_list = random.sample(test_pool, n_test)
This is fold: 01, TestSamples: [23, 18, 10]
Final train loss is -0.4328 with loss_CE of 0.0000 and loss_MinCut of -0.4809, and final test accuracy is 1.0000
This is fold: 02, TestSamples: [16, 32, 1]
Final train loss is -0.3935 with loss_CE of 0.0000 and loss_MinCut of -0.4373, and final test accuracy is 0.6667
This is fold: 03, TestSamples: [26, 22, 0]
Final train loss is -0.4075 with loss_CE of 0.0000 and loss_MinCut of -0.4528, and final test accuracy is 0.0000
This is fold: 04, TestSamples: [19, 8, 29]
Final train loss is -0.4526 with loss_CE of 0.0000 and loss_MinCut of -0.5029, and final test accuracy is 0.6667
This is fold: 05, TestSamples: [5, 20, 12]
Final train loss is -0.4432 with loss_CE of 0.0000 and loss_MinCut of -0.4925, and final test accuracy is 0.6667
This is fold: 06, TestSamples: [13, 9, 4]
Final train loss is -0.4822 with loss_CE of 0.0000 and loss_MinCut of -0.5358, and final test accuracy is 1.0000
This is fold: 07, TestSamples: [2, 14, 11]
Final train loss is -0.4612 with loss_CE of 0.0000 and loss_MinCut of -0.5125, and final test accuracy is 0.6667
This is fold: 08, TestSamples: [24, 28, 7]
Final train loss is -0.4373 with loss_CE of 0.0000 and loss_MinCut of -0.4858, and final test accuracy is 0.3333
This is fold: 09, TestSamples: [17, 31, 21]
Final train loss is -0.4454 with loss_CE of 0.0000 and loss_MinCut of -0.4949, and final test accuracy is 0.6667
This is fold: 10, TestSamples: [6, 15, 27, 25, 33, 3, 30]
Final train loss is -0.4173 with loss_CE of 0.0000 and loss_MinCut of -0.4637, and final test accuracy is 1.0000
2026-01-14 18:01:02
2026-01-14 18:01:02

In [8]:
!Rscript Step3_TCNEnsemble.R

[1] "2026-01-14 21:06:28 CST"
[1] "This is Run1/10"
[1] "This is Run2/10"
[1] "This is Run3/10"
[1] "This is Run4/10"
[1] "This is Run5/10"
[1] "This is Run6/10"
[1] "This is Run7/10"
[1] "This is Run8/10"
[1] "This is Run9/10"
[1] "This is Run10/10"
[1] "2026-01-14 21:06:28 CST"
[1] "This is Image1/34"
[1] "This is Image2/34"
[1] "This is Image3/34"
[1] "This is Image4/34"
[1] "This is Image5/34"
[1] "This is Image6/34"
[1] "This is Image7/34"
[1] "This is Image8/34"
[1] "This is Image9/34"
[1] "This is Image10/34"
[1] "This is Image11/34"
[1] "This is Image12/34"
[1] "This is Image13/34"
[1] "This is Image14/34"
[1] "This is Image15/34"
[1] "This is Image16/34"
[1] "This is Image17/34"
[1] "This is Image18/34"
[1] "This is Image19/34"
[1] "This is Image20/34"
[1] "This is Image21/34"
[1] "This is Image22/34"
[1] "This is Image23/34"
[1] "This is Image24/34"
[1] "This is Image25/34"
[1] "This is Image26/34"
[1] "This is Image27/34"
[1] "This is Image28/34"
[1] "This is Image29/34"
[1]

[1] "2026-01-14 18:01:02 CST"
[1] "This is Run1/10"
[1] "This is Run2/10"
[1] "This is Run3/10"
[1] "This is Run4/10"
[1] "This is Run5/10"
[1] "This is Run6/10"
[1] "This is Run7/10"
[1] "This is Run8/10"
[1] "This is Run9/10"
[1] "This is Run10/10"
[1] "2026-01-14 18:01:03 CST"
[1] "This is Image1/34"
[1] "This is Image2/34"
[1] "This is Image3/34"
[1] "This is Image4/34"
[1] "This is Image5/34"
[1] "This is Image6/34"
[1] "This is Image7/34"
[1] "This is Image8/34"
[1] "This is Image9/34"
[1] "This is Image10/34"
[1] "This is Image11/34"
[1] "This is Image12/34"
[1] "This is Image13/34"
...
[1] "This is Image32/34"
[1] "This is Image33/34"
[1] "This is Image34/34"
[1] "2026-01-14 18:01:21 CST"

In [9]:
!python Step4_ResultVisualization.py

2026-01-14 21:06:48
This is Image1/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image2/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image3/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette li

2026-01-14 18:01:22
This is Image1/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image2/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image3/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image4/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image5/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image6/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image7/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
This is Image8/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
...
This is Image34/34
/home/peiqigao/projects/cyto/CytoCommunity/Tutorial/Supervised/Step4_ResultVisualization.py:123: UserWarning: The palette list has more values (30) than needed (17), which may not be intended.
  CellType_plot = sns.scatterplot(x="x_coordinate", y="y_coordinate", data=target_graph_map, hue="Cell_Type", palette=sns.color_palette("husl", 30), alpha=1.0, s=10.0, legend="full")  # 30 colors at maximum.
2026-01-14 18:02:29

对比了一下调用了GPU的pytorch和没有调用GPU的pytorch运行这个代码（已经简化了的）的时间

之前是16:36:06-18:02:29；大约90min

调用之后就是21:02:18-21:07:55；大约5min